workflow/executor/resource.go

```diff
@@ -80,51 +92,151 @@ func (we *WorkflowExecutor) WaitResource(resourceName string) error {
+			}
+
+			log.Warnf("Waiting for resource %s resulted in non-retryable error %v", resourceName, result.err)
+			return false, result.err
+		})
+
+	close(ctrlChannel)
+
 	if err != nil {
-		return errors.InternalWrapError(err)
-	}
-	reader := bufio.NewReader(stdout)
-	log.Info(strings.Join(cmd.Args, " "))
-	if err := cmd.Start(); err != nil {
-		return errors.InternalWrapError(err)
+		if err == wait.ErrWaitTimeout {
+			log.Warnf("Waiting for resource %s resulted in timeout due to repeated errors", resourceName)
+		} else {
+			log.Warnf("Waiting for resource %s resulted in error %v", err)
+		}
 	}
-	defer func() {
-		_ = cmd.Process.Kill()
-	}()
+
+	return err
+}
+
+func checkResourceState(ctrlChannel <-chan string, resultChannel chan<- waitResult, resourceName string,
+	successReqs labels.Requirements, failReqs labels.Requirements) {
+
 	for {
-		jsonBytes, err := readJSON(reader)
+		ctrlMsg := <-ctrlChannel
+
+		if ctrlMsg != startStr {
+			msg := fmt.Sprintf(
+				"checkResourceState received a non start message '%s' on control Channel, when waiting for resource %s",
+				ctrlMsg, resourceName)
+			log.Warnf(msg)
+			return
+		}
+
+		cmd, reader, err := startKubectlWaitCmd(resourceName)
 		if err != nil {
-			// TODO: it's possible that kubectl might EOF upon an API disconnect.
-			// In this case, we would want to refork kubectl again.
-			return errors.InternalWrapError(err)
+			resultChannel <- waitResult{err: err, result: failStr}
+			return
 		}
-		log.Info(string(jsonBytes))
-		ls := gjsonLabels{json: jsonBytes}
-		for _, req := range failReqs {
-			failed := req.Matches(ls)
-			msg := fmt.Sprintf("failure condition '%s' evaluated %v", req, failed)
-			log.Infof(msg)
-			if failed {
-				// TODO: need a better error code instead of BadRequest
-				return errors.Errorf(errors.CodeBadRequest, msg)
+		defer func() {
+			_ = cmd.Process.Kill()
+		}()
+
+		for {
+			jsonBytes, err := readJSON(reader)
+
+			if err != nil {
+				resultErr := err
+				msg := fmt.Sprintf("Json reader returned error %v. Calling kill (usually superfluous)", err)
+				log.Warnf(msg)
+				// We don't want to write OS specific code so we don't want to call syscall package code. But that means
+				// there is no way to figure out if a process is running or not in an asynchronous manner. exec.Wait will
+				// always block and we need to call that to get the exit code of the process. So we will unconditionally
+				// call exec.Process.Kill and then assume that wait will not block after that. Two things may happen:
+				// 1. Process already exited and kill does nothing (returns error which we ignore) and then we call
+				//    Wait and get the proper return value
+				// 2. Process is running gets, killed with exec.Process.Kill call and Wait returns an error code and we give up
+				//    and don't retry
+				_ = cmd.Process.Kill()
+
+				msg = fmt.Sprintf("Command for kubectl get -w for %s exited. Getting return value using Wait", resourceName)
+				log.Warnf(msg)
```

Comment:
Nit: the pattern of putting msg into a variable and passing msg to log.XXXXf is a little hard to parse. I think it would be fine if we were reusing the message (e.g. in a returned error), but if all we are doing is logging the message, can we just do:
```
log.Warnf("Command for kubectl get -w for %s exited. Getting return value using Wait", resourceName)
```
and save a bunch of lines? Also for code base searches it will be easier to grep for what we are warning/erroring about.

Reply:
Fixed